### ISSUE TITLE
Bring back stack traces in pretty formatter

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber/cucumber-js#624]
As of version 0.8.0 the pretty formatter no longer prints stack traces under failing steps - you have to wait until the end to see why a step failed.

In an ideal world where all scenarios are blazing fast this isn't a problem, but unfortunately I'm not in the ideal world. This means I have to wait a long time before I can start investigating a failure.

Can we have a variant of the pretty formatter (`pretty-verbose` maybe) that spits out errors as they happen?

/cc @tooky @jbpros @mattwynne

**Comment by charlierudolph** [comment]
Related to #553. 

I'm good with this as long as the summary at the end still contains the error so you don't need to scroll back up searching through the output to find the error. That was why I originally moved it to the end. 

I think this could just be an option that gets passed to the formatter. 

```sh
--format-option <key> <value> 
```

I think the options for printing errors as they happen could be nice for the progress formatter too.

Along this line I'd like to move some of our other top level CLI options into something like this since they are just formatter options

```
--snippet-syntax <FILE>
--colors / --no-colors
--backtrace / --no-backtrace
```

**Comment by aslakhellesoy** [comment]
Sounds good!

In Cucumber-JVM I have deprecated `--format` in favour of `--plugin` and it would be great if all impls did this consistently. If you agree, `--plugin-option` would be better

**Comment by charlierudolph** [comment]
I like the idea of `--plugin` but still consider it different for `--format`, and would favor supporting both. Each hook into the cucumber lifecycle, and could use CLI to define options but custom formatters would be passed a writable stream. Currently the cucumber-js framework deals with opening the stream and the formatter closes it in the AfterFeatures event. I want to update it so the framework takes care of closing it as well.

Does cucumber-jvm allow multiple formatters and if so how does it deal with writing to files?

**Comment by charlierudolph** [comment]
@aslakhellesoy bump on my last comment

**Comment by floribon** [comment]
I've found a workaround that works perfectly for me: use the `StepResult` hook and get the error from `stepResult.getFailureException`:

```
this.StepResult(function(stepResult) {
  if (stepResult.getStatus() === 'failed') {
    var exception = stepResult.getFailureException();
    console.error(exception.stack);
  }
});
```

Or do fancier things such as adding colors and indentation:

```
var exception = event.getFailureException();
var stacks = exception.split('\n');
stacks[0] = '\x1b[31m' + stacks[0];
stacks[stacks.length-1] = stacks[stacks.length-1] + '\x1b[0m';
stacks.forEach(function(line) {
  console.error('    ' + line);
});
```

That's not as clean as a built-in option but on the other hand this gives more flexibility to the user.

**Comment by SattaRavi** [comment]
@floribon is there a way i could find which step succeeded or failed from within the step result ?

**Comment by floribon** [comment]
@SattaRavi with version < `2.x` it's in my answer: use `stepResult.getStatus() === 'failed'` to know if that last step failed. If it is the case you can retrieve the reason using `stepResult.getFailureException()`

Since cucumber `2.x`, these are properties directly accessible and not methods anymore: `stepResult.status === 'failed'` and then `stepResult.failureException`

**Comment by SattaRavi** [comment]
thanks @floribon i figured what i was looking for, documenting it here for others finding the thread

using < 2.x 
`this.StepResult(function(stepResult) {
     //step = stepResult.getStep();
             //step.getName()) - gets the step that got executed
    //stepResult.getStatus() - gets the step status passed || failed || skipped
});`

**Comment by charlierudolph** [comment]
Just released 2.2.0 which has a `progress-bar` formatter which I think it the best formatter to use when running many features. It outputs a progress bar and errors as they happen.

**Comment by charlierudolph** [comment]
Just merged in making the error formatting using pretty output. The pretty formatter has thus been deleted. You can use the `progress-bar` for essentially this.
